### PR TITLE
[RFC-191]: EVM-740 - Move distribute rewards to first block in epoch

### PR DESCRIPTION
**Pull request by goran-ethernal**
# Description

`ValidatorSet` contract on supernet is used as a wrapper around staked tokens by validators on `L2`. For governance, it is used as a voting token as well, which makes sense, since we are only allowing validators to propose and vote for governance proposals.

But, to be able to use `ValidatorSet` as vote token in governance, we needed to change it to implement `ERC20VotesUpgradable` implementation, which changed how the contract checkpoints its total supply at a certain point in time, and how it returns `totalSupplyAt`, which is used by `RewardPool` contract to accurately distribute rewards.
Now `totalSupplyAt` has a check which prevents calling it for the last checkpointed total supply, if they happen in the same block, and because of that, reward distribution failed.

This PR moves reward distribution to the first block of epoch. So when an epoch completes, and gets committed (which will checkpoint total supply at `ValidatorSet` contract), the new epoch will start, and in the first block of new epoch, proposer will add the distribute rewards transaction for the previously completed epoch.

This has three benefits:
1. It lowers gas consumption on epoch ending blocks.
2. It enables more accurate reward distribution.
3. We can use `ValidatorSet` as voting token for governance.

**IMPORTANT NOTE**:
Rewards will only be distributed in the first block of an epoch if `governance` fork is enabled. If it is not enabled, rewards will be distributed old way, at epoch ending block. 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### CODE REVIEW

File: consensus/polybft/consensus_runtime.go [hunk 1]
```diff
@@ -490,66 +496,105 @@ func (c *consensusRuntime) restartEpoch(header *types.Header) (*epochMetadata, e
...
-	currentBlock *types.Header,
-	epoch *epochMetadata,
-) (*contractsapi.CommitEpochValidatorSetFn,
-	*contractsapi.DistributeRewardForRewardPoolFn, error) {
-	uptimeCounter := map[types.Address]int64{}
-	blockHeader := currentBlock
-	epochID := epoch.Number
-	totalBlocks := int64(0)
+// calculateCommitEpochInput calculates commit epoch input data
+func calculateCommitEpochInput(
```
Comment (igorcrevar) [inline consensus/polybft/consensus_runtime.go]: Maybe we should rename this helper function to something like `getCommitEpochInput` or similar and make that function outside of `consensusRuntime`. There is no calculation, so name is misleading

Comment (goran-ethernal) [reply consensus/polybft/consensus_runtime.go]: Done: https://github.com/0xPolygon/polygon-edge/pull/1735/commits/4ebb065d1692b278a0013f336380472289bb324d

File: consensus/polybft/governance_fork.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	oldRewardLookbackSize = uint64(2) // number of blocks to calculate commit epoch info from the previous epoch
+	newRewardLookbackSize = uint64(1)
+)
+
+// shouldAddOrValidateRewardDistribution indicates if reward distribution transaction
+// should be validated or added at given block
+// if governance fork is enabled, reward distribution is only present on the first block of epoch
+// and if we are not at the start of chain
+// if governance fork is not enabled, reward distribution is only present at the epoch ending block
+func shouldAddOrValidateRewardDistribution(isFirstBlockOfEpoch, isEndOfEpoch bool,
```
Comment (Stefan-Ethernal) [inline consensus/polybft/governance_fork.go]: WDYT?
```suggestion
func isRewardDistributionBlock(isFirstBlockOfEpoch, isEndOfEpoch bool,
```

File: consensus/polybft/consensus_runtime_test.go [hunk 3]
```diff
@@ -389,11 +397,10 @@ func TestConsensusRuntime_FSM_EndOfEpoch_BuildCommitEpoch(t *testing.T) {
 	validatorAccounts := validator.NewTestValidatorsWithAliases(t, []string{"A", "B", "C", "D", "E", "F"})
 	validators := validatorAccounts.GetPublicIdentities()
 
-	lastBuiltBlock, headerMap := createTestBlocks(t, 9, epochSize, validators)
+	lastBuiltBlock := &types.Header{Number: 9}
```
Comment (Stefan-Ethernal) [inline consensus/polybft/consensus_runtime_test.go]: Nitpick, we can spare this line, by inlining it when instantiating `consensusRuntime` :wink:

File: consensus/polybft/fsm.go [hunk 4]
```diff
@@ -36,11 +36,11 @@ var (
...
 	errDistributeRewardsTxDoesNotExist = errors.New("distribute rewards transaction is " +
-		"not found in the epoch ending block")
-	errDistributeRewardsTxNotExpected = errors.New("didn't expect distribute rewards transaction " +
-		"in a non epoch ending block")
+		"not found in the given block, but should be added")
+	errDistributeRewardsTxNotExpected = errors.New("distribute rewards transaction " +
+		"is not expected at this block")
 	errDistributeRewardsTxSingleExpected = errors.New("only one distribute rewards transaction is " +
-		"allowed in an epoch ending block")
+		"allowed at the expected block")
```
Comment (Stefan-Ethernal) [inline consensus/polybft/fsm.go]:
```suggestion
		"allowed in the given block")
```

File: consensus/polybft/fsm.go [hunk 5]
```diff
@@ -36,11 +36,11 @@ var (
 	errCommitEpochTxSingleExpected = errors.New("only one commit epoch transaction is allowed " +
 		"in an epoch ending block")
 	errDistributeRewardsTxDoesNotExist = errors.New("distribute rewards transaction is " +
-		"not found in the epoch ending block")
-	errDistributeRewardsTxNotExpected = errors.New("didn't expect distribute rewards transaction " +
-		"in a non epoch ending block")
+		"not found in the given block, but should be added")
```
Comment (Stefan-Ethernal) [inline consensus/polybft/fsm.go]:
```suggestion
		"not found in the given block, though it is expected to be present")
```

**Review by Stefan-Ethernal — APPROVED** [review]
Overall looks good!

File: consensus/polybft/governance_fork.go (hunk 2 above)
Comment (goran-ethernal) [reply consensus/polybft/governance_fork.go]: Done: https://github.com/0xPolygon/polygon-edge/pull/1735/commits/7fa9afd2985856e7b552d15dc83aa367daa78411

File: consensus/polybft/fsm.go (hunk 4 above)
Comment (goran-ethernal) [reply consensus/polybft/fsm.go]: Done: https://github.com/0xPolygon/polygon-edge/pull/1735/commits/7fa9afd2985856e7b552d15dc83aa367daa78411

File: consensus/polybft/consensus_runtime_test.go (hunk 3 above)
Comment (goran-ethernal) [reply consensus/polybft/consensus_runtime_test.go]: Done: https://github.com/0xPolygon/polygon-edge/pull/1735/commits/7fa9afd2985856e7b552d15dc83aa367daa78411

File: consensus/polybft/fsm.go (hunk 5 above)
Comment (goran-ethernal) [reply consensus/polybft/fsm.go]: Done: https://github.com/0xPolygon/polygon-edge/pull/1735/commits/7fa9afd2985856e7b552d15dc83aa367daa78411

**Review by igorcrevar — APPROVED** [review]

**Pull request merged**